### PR TITLE
Check runnable selected resource scope before finding matches

### LINKED ISSUE #575: Cannot list resource "pipelines" in API group "tekton.dev" at the cluster scope
### Bug description:
Runnable seems to be missing permission to list tekton pipelines, even when that permission is granted to the service account in the developer namespace.

### Steps to reproduce:
It seems to happen when you submit a workload before the associated tekton pipeline exists in the cluster.

### Expected behavior:
The order of resources submitted to the cluster shouldn't break things.

### Actual behavior:
It breaks.

### Logs, k8s object dumps:
```
kubectl describe runnables -A                      
Name:         spring-petclinic
Namespace:    my-apps
Labels:       carto.run/cluster-supply-chain-name=source-test-scan-to-url
              carto.run/cluster-template-name=testing-pipeline
              carto.run/resource-name=source-tester
              carto.run/template-kind=ClusterSourceTemplate
              carto.run/workload-name=spring-petclinic
              carto.run/workload-namespace=my-apps
Annotations:  <none>
API Version:  carto.run/v1alpha1
Kind:         Runnable
Metadata:
  Creation Timestamp:  2022-02-01T12:43:20Z
  Generation:          1
  Owner References:
    API Version:           carto.run/v1alpha1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Workload
    Name:                  spring-petclinic
    UID:                   f42cc204-e02b-4daa-8f21-9d3535746206
  Resource Version:        81361018
  UID:                     0b61a711-6c08-4918-8186-610d7d6c2023
Spec:
  Inputs:
    Source - Revision:  main/5d6aae6e09e91658f1f086a2c61f434ae55fe974
    Source - URL:       http://source-controller.flux-system.svc.cluster.local./gitrepository/my-apps/spring-petclinic/5d6aae6e09e91658f1f086a2c61f434ae55fe974.tar.gz
  Run Template Ref:
    Kind:  ClusterRunTemplate
    Name:  tekton-source-pipelinerun
  Selector:
    Matching Labels:
      apps.tanzu.vmware.com/pipeline:  test
    Resource:
      API Version:  tekton.dev/v1beta1
      Kind:         Pipeline
Status:
  Conditions:
    Last Transition Time:  2022-02-01T12:43:24Z
    Message:               unable to resolve selector [map[apps.tanzu.vmware.com/pipeline:test]], apiVersion [tekton.dev/v1beta1], kind [Pipeline]: failed to list objects matching selector [map[apps.tanzu.vmware.com/pipeline:test]]: unable to list from api server: pipelines.tekton.dev is forbidden: User "system:serviceaccount:my-apps:default" cannot list resource "pipelines" in API group "tekton.dev" at the cluster scope
    Reason:                TemplateStampFailure
    Status:                False
    Type:                  RunTemplateReady
    Last Transition Time:  2022-02-01T12:43:24Z
    Message:               unable to resolve selector [map[apps.tanzu.vmware.com/pipeline:test]], apiVersion [tekton.dev/v1beta1], kind [Pipeline]: failed to list objects matching selector [map[apps.tanzu.vmware.com/pipeline:test]]: unable to list from api server: pipelines.tekton.dev is forbidden: User "system:serviceaccount:my-apps:default" cannot list resource "pipelines" in API group "tekton.dev" at the cluster scope
    Reason:                TemplateStampFailure
    Status:                False
    Type:                  Ready
  Observed Generation:     1
Events:                    <none>
```

### PR DESCRIPTION
**Pull request by martyspiewak**
[PR requirements]: https://github.com/vmware-tanzu/cartographer/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines

## Changes proposed by this PR

Runnable checks the scope of the selected resource kind before attempting to find matches. This way it will only ever make one `ListUnstructured` call- either in the namespace or at the cluster scope.

closes #575 

## Release Note

* Fixes bug where Runnable would return a misleading error saying the service account doesn't have permission to list the selected object at the cluster scope (#575)

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [ ] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [ ] Removed non-atomic or `wip` commits
- [ ] Filled in the [Release Note](#Release-Note) section above 
- [ ] Modified the docs to match changes

### CODE REVIEW

File: pkg/realizer/runnable/realizer.go [hunk 1]
```diff
@@ -147,40 +149,39 @@ func (p *runnableRealizer) Realize(ctx context.Context, runnable *v1alpha1.Runna
...
-func resolveSelector(ctx context.Context, selector *v1alpha1.ResourceSelector, repository repository.Repository, namespace string) (map[string]interface{}, error) {
+func (r *runnableRealizer) resolveSelector(ctx context.Context, selector *v1alpha1.ResourceSelector, repository repository.Repository, discoveryClient discovery.DiscoveryInterface, namespace string) (map[string]interface{}, error) {
 	log := logr.FromContextOrDiscard(ctx)
 
 	if selector == nil {
 		return nil, nil
 	}
 
-	results, err := repository.ListUnstructured(ctx, schema.FromAPIVersionAndKind(selector.Resource.APIVersion, selector.Resource.Kind), namespace, selector.MatchingLabels)
+	apiResourceList, err := discoveryClient.ServerResourcesForGroupVersion(selector.Resource.APIVersion)
```
Comment (cirocosta) [inline pkg/realizer/runnable/realizer.go:159]: do you know if there's any impact performance-wise with having this being run on every `resolveSelector` call? I'm not very familiar with the internals of the discovery client, but, IIRC, `kubectl` makes use of a cached version of it that persists results on disk to make things fast 🤔

File: pkg/realizer/client/client.go [hunk 2]
```diff
@@ -18,25 +18,34 @@ import (
...
 		cl, err := client.New(config, client.Options{})
 		if err != nil {
-			return nil, fmt.Errorf("creating client: %w", err)
+			return nil, nil, fmt.Errorf("creating client: %w", err)
 		}
 
-		return cl, nil
+		var discoveryClient discovery.DiscoveryInterface
+		if needDiscovery {
+			discoveryClient, err = discovery.NewDiscoveryClientForConfig(restConfig)
```
Comment (cirocosta) [inline pkg/realizer/client/client.go]: related to https://github.com/vmware-tanzu/cartographer/pull/660/files#r813367165, have you considered [`NewMemCacheClient`](https://github.com/kubernetes/client-go/blob/41e0447e772813a0942f06f64f9eaf07a70cbdf8/discovery/cached/memory/memcache.go#L223-L233)?

File: pkg/realizer/runnable/realizer.go [hunk 3]
```diff
@@ -147,40 +149,39 @@ func (p *runnableRealizer) Realize(ctx context.Context, runnable *v1alpha1.Runna
...
-	if len(results) == 0 {
-		log.V(logger.DEBUG).Info("selector did not match any objects, checking resolveClusterScopedSelector")
-		return resolveClusterScopedSelector(ctx, selector, repository)
-	} else if len(results) > 1 {
-		log.V(logger.DEBUG).Info("selector matched multiple objects")
-		return nil, fmt.Errorf("selector matched multiple objects")
+	var namespaced bool
+	for _, apiResource := range apiResourceList.APIResources {
+		if apiResource.Kind == selector.Resource.Kind {
+			namespaced = apiResource.Namespaced
```
Comment (cirocosta) [inline pkg/realizer/runnable/realizer.go:168]: should we break out of the loop given that we've found that it's namespaced?

Comment (martyspiewak) [reply pkg/realizer/runnable/realizer.go:168]: good catch- missed that when I was refactoring

File: pkg/realizer/client/client.go (hunk 2 above)
Comment (martyspiewak) [reply pkg/realizer/client/client.go]: I had not seen this- thanks! switched to using this

File: pkg/realizer/runnable/realizer.go (hunk 1 above)
Comment (martyspiewak) [reply pkg/realizer/runnable/realizer.go:159]: switched to the cached client, as you suggested below.

**Review by cirocosta — APPROVED** [review]

**Pull request merged**